--- linux-3.2.44/net/ipv4/Makefile.orig	2013-05-08 20:43:50.892913537 +0300
+++ linux-3.2.44/net/ipv4/Makefile	2013-05-08 20:43:59.853005836 +0300
@@ -20,6 +20,7 @@
 obj-$(CONFIG_NET_IPIP) += ipip.o
 obj-$(CONFIG_NET_IPGRE_DEMUX) += gre.o
 obj-$(CONFIG_NET_IPGRE) += ip_gre.o
+obj-$(CONFIG_NET_EOIP) += eoip.o
 obj-$(CONFIG_SYN_COOKIES) += syncookies.o
 obj-$(CONFIG_INET_AH) += ah4.o
 obj-$(CONFIG_INET_ESP) += esp4.o
--- linux-3.2.44/net/ipv4/Kconfig.orig	2013-05-08 20:42:23.736015745 +0300
+++ linux-3.2.44/net/ipv4/Kconfig	2013-05-08 20:43:26.652663869 +0300
@@ -211,6 +211,12 @@
 	  Network), but can be distributed all over the Internet. If you want
 	  to do that, say Y here and to "IP multicast routing" below.
 
+config NET_EOIP
+	tristate "IP: EOIP tunnels over IP"
+	depends on (IPV6 || IPV6=n) && NET_IPGRE_DEMUX
+	help
+	  Tunneling ethernet over IP. Compatible with MikroTik
+
 config IP_MROUTE
 	bool "IP: multicast routing"
 	depends on IP_MULTICAST
